payjoin-cli/src/db/v2.rs

```diff
@@ -1,21 +1,38 @@
+use std::sync::Arc;
+
 use bitcoincore_rpc::jsonrpc::serde_json;
-use payjoin::receive::v2::Receiver;
+use payjoin::bitcoin::hex::DisplayHex;
+use payjoin::persist::{Persister, Value};
+use payjoin::receive::v2::{Receiver, ReceiverToken};
 use payjoin::send::v2::Sender;
-use sled::{IVec, Tree};
+use sled::Tree;
 use url::Url;
 
 use super::*;
 
-impl Database {
-    pub(crate) fn insert_recv_session(&self, session: Receiver) -> Result<()> {
-        let recv_tree = self.0.open_tree("recv_sessions")?;
-        let key = &session.id();
-        let value = serde_json::to_string(&session).map_err(Error::Serialize)?;
-        recv_tree.insert(key.as_slice(), IVec::from(value.as_str()))?;
+#[derive(Clone)]
+pub(crate) struct ReceiverPersister(pub Arc<Database>);
+impl Persister<Receiver> for ReceiverPersister {
+    type Token = ReceiverToken;
+    type Error = crate::db::error::Error;
+    fn save(&mut self, value: Receiver) -> std::result::Result<ReceiverToken, Self::Error> {
+        let recv_tree = self.0 .0.open_tree("recv_sessions")?;
+        let key = value.key();
+        let value = serde_json::to_vec(&value).map_err(Error::Serialize)?;
+        recv_tree.insert(key.clone(), value.as_slice())?;
         recv_tree.flush()?;
-        Ok(())
+        Ok(key)
     }
+    fn load(&self, key: ReceiverToken) -> std::result::Result<Receiver, Self::Error> {
+        let recv_tree = self.0 .0.open_tree("recv_sessions")?;
+        let value = recv_tree.get(key.as_ref())?.ok_or(Error::NotFound(
+            key.as_ref().to_hex_string(payjoin::bitcoin::hex::Case::Lower),
```

Comment:
Wouldn't it be more appropriate for `key` to be `Display` for this application?

Seems it needs to be `AsRef` for `get`, but `Display` for errors

Reply:
Yes key not impls bounded by Display